[OTX] Add SupCon recipes for classification task

arrufat opened the pull request:
This PR contains recipes for training models using a [SupCon](https://arxiv.org/abs/2004.11362)-like style.

These recipes aim to increase the current classification baseline when the data is scarce.

Procedure:
- Train a model with two heads:
  - classification head with standard cross-entropy loss.
  - SupCon head with custom SelfSL loss (currently it uses the [Barlow Twins](https://arxiv.org/abs/2103.03230) loss).
- At inference, the SupCon head is discarded, hence incurring no overhead at runtime.

The recipes include:
- a data pipeline using TwoCrop transform (to generate two views of an image for the SelfSL head)
- a SupConClassifier (in charge of handling the two views of images)
- a SupConClsHead (the one that uses the dual-head approach)
- a SelfSL loss (the Barlow Twins loss)

So far, this method will be used in OTX when specifying the `TrainType` as `SUPERVISEDCONTRASTIVE`.

Commit:
Add SupCon recipes for multilabel classification

Commit:
Update batch size in two crop data

Commit:
Refactor the BarlowTwinsLoss for extensibility

jaegukhyun commented on recipes/stages/classification/supcon.yaml:
```diff
@@ -0,0 +1,21 @@
+
+evaluation:
+    metric: ['accuracy', 'class_accuracy']
+
```
Could you explain why task adapt op disappear in this recipe?f

arrufat replied:
Task adapt replaces whichever loss function I specify in the model with CrossEntropyLoss.
And I think it also changed the number of classes or something.

It wasn't working before, but maybe after the refactoring I pushed this morning, it might work.

I will check and revert it if it works.

arrufat replied:
Yes, I've run a couple of tests, and it works now with `task_adapt` after the refactoring. Thank you!

jaegukhyun reviewed (commented):
Genreally, LGTM. I left one comment

Commit:
Add task_adapt to SupCon recipe

jaegukhyun commented on mpa/modules/models/heads/supcon_cls_head.py:
```diff
@@ -0,0 +1,94 @@
+        if isinstance(topk, int):
+            topk = (topk,)
+        for _topk in topk:
+            assert _topk > 0, "Top-k should be larger than 0"
```
Is it okay if _topk is larger than num_classes?

arrufat replied:
Fixed!

jaegukhyun reviewed (commented):
I left one comment, and I have a question. Where is the part for dividing learning rate by 2? I heard that is necessary because of changes in batch size

Commit:
Fix bug in topk when it's larger than num_classes

Also rename the linear member to fc and fc_feats to cls_score for
better alignment with the rest of the code.

arrufat commented:
> I left one comment, and I have a question. Where is the part for dividing learning rate by 2? I heard that is necessary because of changes in batch size

As discussed with Wonju, for now, this logic is not implemented, we thought it would be better to do it manually and, once it because the default method for small datasets, add the logic in the code. But it's up for discussion.

Commit:
Remove extract_prob method to align with SAMImageClassifier

Commit:
Simplify multilabel and hierarchical kwargs logic

jaegukhyun reviewed (approved):
LGTM, @harimkang, @supersoob, @JihwanEom anyone, could you check this PR once?

harimkang reviewed (approved):
If this works well, there seems to be no structural problem. LGTM.
